File: pose_estimation/kapao/detect.py
import cv2
import argparse
import numpy as np
import onnxruntime as ort

config = {'person_conf_thres': 0.55, 'person_iou_thres': 0.45,
          'kp_conf_thres': 0.3, 'kp_iou_thres': 0.40, 'conf_thres_kp_person': 0.3,
           # if the center point of kp_bbox is within the range of 'overwrite_tol' pixel, it will replace the predict point
          'overwrite_tol': 12.5,
          'kp_face': [0, 1, 2, 3, 4], 'use_kp_dets': True,
          'segments': {1: [5, 6], 2: [5, 11], 3: [11, 12], 4: [12, 6],
                       5: [5, 7], 6: [7, 9], 7: [6, 8], 8: [8, 10],
                       9: [11, 13], 10: [13, 15], 11: [12, 14], 12: [14, 16], },
          # 'crowd_segment': {1:[0, 13], 2:[1, 13], 3:[0, 2], 4:[12, 6], 5: [5, 7],
          #                   6:[7, 9], 7:[6, 8], 8:[8, 10], },
          # 'crowd_kp_face': [],
          'input_size': (640, 640),
          'anchors': [[10, 13, 16, 30, 33, 23],  # P3/8
                      [30, 61, 62, 45, 59, 119],  # P4/16
                      [116, 90, 156, 198, 373, 326]],  # P5/32
          'stride': [8., 16., 32.]  # grid size
          }


class kapao():  # Keypoint and Pose as Object
    def __init__(self, model_path):
        with open('./cpp_detect/class.names', 'rt') as fd:
            self.classes = fd.read().rstrip('\n').split('\n')
        self.lines = config['segments']
        self.kp_face = config['kp_face']

        self.num_classes = len(self.classes)
        self.input_height, self.input_width = config['input_size']
        anchors = config['anchors']
        self.stride = np.array(config['stride'])
        self.nl = len(anchors)  # number of heads
        self.na = len(anchors[0]) // 2  # number of anchors per head
        self.grid = [np.zeros(1)] * self.nl
        self.anchor_grid = np.asarray(anchors, dtype=np.float32).reshape(self.nl, -1, 2)
        # self.net = cv2.dnn.readNetFromONNX(model_path)  # TODO: whatt?
        self.net = cv2.dnn.readNet(model_path)  # TODO: whatt?
        self._input_names = 'images'
        self.last_ind = 5 + self.num_classes  # total index: 5(x,y,w,h,conf) + num_classes + 2*num_kpts

    def resize_image(self, src_img, keep_ratio=True, dynamic=False):
        top, left, h_new, w_new = 0, 0, self.input_height, self.input_width
        if keep_ratio and src_img.shape[0] != src_img.shape[1]:
            hw_scale = src_img.shape[0] / src_img.shape[1]  # h / w
            if hw_scale > 1:
                h_new, w_new = self.input_height, int(self.input_width / hw_scale)
                img = cv2.resize(src_img, (w_new, h_new), interpolation=cv2.INTER_AREA)
                if not dynamic:
                    left = int((self.input_width - w_new) * 0.5)
                    img = cv2.copyMakeBorder(img, 0, 0, left, self.input_width - w_new,
                                             cv2.BORDER_CONSTANT, value=(114, 114, 114))
            else:
                h_new, w_new = int(self.input_height * hw_scale), self.input_width
                img = cv2.resize(src_img, (w_new, h_new), interpolation=cv2.INTER_AREA)
                if not dynamic:
                    top = int((self.input_height - h_new) * 0.5)
                    img = cv2.copyMakeBorder(img, top, self.input_height - h_new - top, 0, 0,
                                             cv2.BORDER_CONSTANT, value=(114, 114, 114))
        else:
            img = cv2.resize(src_img, (self.input_width, self.input_height), inperpolation=cv2.INTER_AREA)
        return img, h_new, w_new, top, left

    def _make_grid(self, nx=20, ny=20):
        xv, yv = np.meshgrid(np.arange(ny), np.arange(nx))
        return np.stack((xv, yv), 2).reshape((-1, 2)).astype(np.float32)

    def preprocess(self, img):
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        img = img.astype(np.float32) / 255.0
        return img

    def postprocess(self, frame, outs, pad_size=None):
        """"""
        frame_height, frame_width = frame.shape[:2]
        h_new, w_new, h_pad, w_pad = pad_size
        h_ratio, w_ratio = frame_height / h_new, frame_width / w_new

        # 1. Scan through all the bounding boxes output from the network and keep only the
        # ones with high confidence scores. Assign the box's class label as the class with the highest score.
        person_confidences, kp_confidences = [], []  # confidences
        person_boxes, kp_boxes = [], []  # bboxes
        person_classsIds, kp_classIds = [], []  # class indices
        person_rowins = []
        for i in range(outs.shape[0]):
            detection = outs[i, :]
            x, y, w, h, obj_conf = detection[:5]
            scores = detection[5: self.last_ind]
            classId = np.argmax(scores)
            confidence = scores[classId] * obj_conf  # class_scores * object_confidence
            if classId == 0:
                if obj_conf > config['person_conf_thres'] and confidence > config['person_conf_thres']:
                    center_x = int((x - w_pad) * w_ratio)
                    center_y = int((y - h_pad) * h_ratio)
                    width = int(w * w_ratio)
                    height = int(h * h_ratio)
                    left = max(int(center_x - width * 0.5), 0)
                    top = max(int(center_y - height * 0.5), 0)

                    person_confidences.append(float(confidence))
                    person_boxes.append([left, top, width, height])
                    person_classsIds.append(classId)
                    person_rowins.append(i)
            else:
                if obj_conf > config['kp_conf_thres'] and confidence > config['kp_conf_thres']:

                    center_x = int((x - w_pad) * w_ratio)
                    center_y = int((y - h_pad) * h_ratio)
                    width = int(w * w_ratio)
                    height = int(h * h_ratio)
                    left = max(int(center_x - width * 0.5), 0)
                    top = max(int(center_y - height * 0.5), 0)

                    kp_confidences.append(confidence)
                    kp_boxes.append([left, top, width, height])
                    kp_classIds.append(classId)

        # 2. Perform non-maximum suppression to eliminate redundant overlapping boxes with lower confidences.
        person_indices = cv2.dnn.NMSBoxes(person_boxes, person_confidences, config['person_conf_thres'],
                                          config['person_iou_thres']).flatten()
        kp_indices = cv2.dnn.NMSBoxes(kp_boxes, kp_confidences, config['kp_conf_thres'],
                                      config['kp_iou_thres']).flatten()
        poses = []  # keypoints
        for i in person_indices:
            if person_confidences[i] > config['conf_thres_kp_person']:
                pose = outs[person_rowins[i], self.last_ind:].reshape((-1, 2))  # [num_joints, 2]
                pose[:, 0] = (pose[:, 0] - w_pad) * w_ratio
                pose[:, 1] = (pose[:, 1] - h_pad) * h_ratio
                poses.append(pose)
        nd = len(poses)
        poses = np.array(poses)  # [nd, num_joints, 2]
        poses = np.concatenate((poses, np.zeros((nd, poses.shape[1], 1))), axis=-1)  # [nd, num_joints, 3]
        # fuse keypoints bbox and keypoints
        for j in kp_indices:
            box = kp_boxes[j]
            x = box[0] + 0.5 * box[2]
            y = box[1] + 0.5 * box[3]
            pt_id = kp_classIds[j] - 1  # 1~num_joints -> 0 ~ num_joints - 1
            pose_kps = poses[:, pt_id, :]
            dist = np.linalg.norm(pose_kps[:, :2] - np.array([[x, y]]), axis=-1)
            kp_match = np.argmin(dist)
            # TODO: why we need to use pose_kps[kp_match, 2], which is always zero 0.
            if kp_confidences[j] > pose_kps[kp_match, 2] and dist[kp_match] < config['overwrite_tol']:
                poses[kp_match, pt_id, :] = np.array([x, y, kp_confidences[j]])

        # 3. Visualization: draw line and bbox
        for i in person_indices:
            left, top, width, height = person_boxes[i]
            frame = self.drawPred(frame, person_classsIds[i], person_confidences[i],
                                  left, top, left + width, top + height)
        for pose in poses:
            for seg in self.lines.values():
                pt1 = (int(pose[seg[0], 0]), int(pose[seg[0], 1]))
                pt2 = (int(pose[seg[1], 0]), int(pose[seg[1], 1]))
                cv2.line(frame, pt1, pt2, (255, 0, 255), 1)
            for x, y, c in pose:
                if c > 0:
                    cv2.circle(frame, (int(x), int(y)), 1, (0, 0, 255), 1)
            # TODO: add args to control whether to draw face keypoints
            # for x, y, c, in pose[self.kp_face]:
            #     cv2.circle(frame, (int(x), int(y)), 1, (255, 0, 255), 1)
        # TODO: add args to control weather to draw keypoint bounding boxes
        # for i in kp_indices:
        #     left, top, width, height = kp_boxes[i]
        #     frame = self.drawPred(frame, kp_classIds[i], kp_confidences[i],
        #                           left, top, left+width, top+height)
        return frame

    def drawPred(self, frame, classId, conf, x1, y1, x2, y2):
        # 1. Draw a bounding box. (x_left, y_top, x_right, y_bottom) = (x1, y1, x2, y2)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 0, 255), thickness=1)
        label = '%.2f' % conf
        label = '%s:%s' % (self.classes[classId], label)

        # 2. Display the label at the top of the bounding box
        labelSize, baseLine = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
        y1 = max(y1, labelSize[1])  # top
        # cv2.rectangle(frame, (x1, y1-round(1.5*labelSize[1])), (x1+round(1.5*labelSize[0]), y1+baseLine),
        #               (255, 255, 255), cv2.FILLED)
        cv2.putText(frame, label, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), thickness=1)
        return frame

    def detect(self, src_img):
        img, h_new, w_new, h_pad, w_pad = self.resize_image(src_img)
        blob = cv2.dnn.blobFromImage(img, scalefactor=1 / 255.0, swapRB=True)
        # blob = cv2.dnn.blobFromImage(self.preprocess(img))
        # 1. Set the input to the network
        self.net.setInput(blob, self._input_names)

        # 2. Runs the forward pass to get output of the output layers
        # TODO: check the dimensions of outs [na*(w1*h1+w2*h2+w3*h3), 6 + 3*num_kp]
        # outs = self.net.forward(self.net.getUnconnectedOutLayersNames())[0].squeeze(axis=0)
        outs = self.net.forward()  # [25200, 57]
        # outs = 1 / (1 + np.exp(-outs))  # sigmoid

        # 3. Decode inference output
        ig = 0  # the index of beginning grid
        # outs = np.zeros_like(outs)
        # outs[0, :6] = [0, 0, 0.5, 0.5, 1, 1]
        # outs[23000, :6] = [0, 0, 0.5, 0.5, 1, 1]
        # outs[2300, :7] = [0, 0, 0.5, 0.5, 1, 0, 1]
        for i in range(len(outs)):
            # if outs[i, 4] > config['person_conf_thres']:
            with open("./detect_results.txt", 'a+') as fd:
                line = list(outs[i, :])
                line = ["%.3f" % x for x in line]
                line = " ".join(line) + ' '
                fd.write(line)

        for i in range(self.nl):
            h, w = int(self.input_height / self.stride[i]), int(self.input_width / self.stride[i])
            length = int(self.na * h * w)  # the number of grids in an output layer
            if self.grid[i].shape[2:4] != (h, w):
                self.grid[i] = self._make_grid(w, h)
            # In COCO dim-1: [tx, ty, tw, th, conf, c_person, c1, c2, ...,c17, x1, y1, ..., x17, y17]
            xy = outs[ig:ig + length, 0:2]  # the coordinates of bounding box center
            wh = outs[ig:ig + length, 2:4]  # the width and height of bounding box
            kp = outs[ig:ig + length, self.last_ind:]  # the coordinates (x1, y1, .., xn, yn) of keypoints
            # # TODO: check this, because of not using sigmoid to decode outputs
            # yolo.py has done the post-process to decode the output
            xy = int(self.stride[i]) * (2. * xy - 0.5 + np.tile(self.grid[i], (self.na, 1)))
            wh = (2. * wh) ** 2 * np.repeat(self.anchor_grid[i], h * w, axis=0)
            num_kp = (outs.shape[1] - self.last_ind) // 2  # len([x1, y1, ..., xn, yn])  // 2
            kp = 4. * kp - 2.
            kp *= np.tile(np.repeat(self.anchor_grid[i], h * w, axis=0), (1, num_kp))
            kp += np.tile(np.tile(self.grid[i], (self.na, 1)) * int(self.stride[i]), (1, num_kp))

            outs[ig:ig + length, 0:2] = xy
            outs[ig:ig + length, 2:4] = wh
            outs[ig:ig + length, self.last_ind:] = kp

            ig += length
        src_img = self.postprocess(src_img, outs, pad_size=(h_new, w_new, h_pad, w_pad))
        return src_img


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    # parser.add_argument('--img-path', type=str, default='data/data/coco/images/test2017/000000000057.jpg',
    parser.add_argument('--img-path', type=str,
                        # default='data/data/coco/images/test2017/000000000057.jpg',
                        default="/home/huangzhiyong/Project/kapao/res/crowdpose_100024.jpg",
                        help='image path')
    parser.add_argument('--model-path', type=str, default='./cpp_detect/model.onnx')
    args = parser.parse_args()

    print(f"{args.model_path}")
    # session = ort.InferenceSession(args.model_path)
    # img0 = cv2.imread(args.img_path)
    # img1 = cv2.resize(img0, (640, 640))
    # ori_image = cv2.cvtColor(img1, cv2.COLOR_BGR2RGB)
    # ori_image = ori_image.astype(np.float32) / 255.0
    # img_input = np.transpose(ori_image, [2, 0, 1])[None]
    # results = session.run(['output'], {'images': img_input})[0]
    # print(f"{type(results)=}")
    # print(f"{results.shape=}")
    model = kapao(args.model_path)
    img_input = cv2.imread(args.img_path)
    print(f"{img_input.shape=}")
    img_input = model.detect(img_input)
    print(f"{img_input.shape=}")
    # img_input = model.drawPred(img_input, 0, 1, 300, 300, 350, 350)

    window_name = "Keypoint and Pose as Object (KAPAO)"
    cv2.namedWindow(window_name, 0)
    cv2.imshow(window_name, img_input)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: pose/kapao_with_kp_conf/cpp_detect/detect.cpp
#include <fstream>
#include <sstream>
#include <iostream>
#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

//#include <onnxruntime_cxx_api.h>

using namespace cv;
using namespace dnn;
using namespace std;

struct NetConfig
{
    float person_conf_thres;
    float person_iou_thres;
    float kp_conf_thres;
    float kp_iou_thres;
    float conf_thres_kp_person;
    int overwrite_tal;
    bool use_kp_dets;
};

int endswith(string s, string sub)
{
    return s.rfind(sub) == (s.length() - sub.length()) ? 1 : 0;
}

int kp_face[5] = {0, 1, 2, 3, 4};
int segment[12][12] = {{5,6}, {5, 11}, {11, 12}, {12, 6}, {5, 7}, {7, 9},
                        {6, 8}, {8, 10},{11, 13}, {13, 15}, {12, 14}, {14, 16}};

class KAPAO  // model KAPAO 
{
    public:
        KAPAO(NetConfig config, string model_path);
        void detect(Mat& frame);
    private:
        const int width = 640;  // input image width
        const int height = 640; // input image height
        const int num_stride = 3;
        const bool keep_ratio = true;
        vector<string> class_names;
        int num_class;
        int num_lines;
        int num_face_pts;
        int* plines;

        Net net;
        NetConfig config;
        void drawPred(float conf, int left, int top, int right, int bottom, Mat& frame, int classid);
        Mat resize_image(Mat src_img, int* h_new, int* w_new, int* top, int* left);
        const float anchors[3][6] = {{10, 13, 16,30, 33,23}, {30, 61, 62, 45, 59, 119}, {116,90, 156, 198, 373, 326}};
};

KAPAO::KAPAO(NetConfig config, string model_path)
{
    this->config.person_conf_thres = config.person_conf_thres;
    this->config.person_iou_thres = config.person_iou_thres;
    this->config.kp_conf_thres = config.kp_conf_thres;
    this->config.kp_iou_thres = config.kp_iou_thres;
    this->config.conf_thres_kp_person = config.conf_thres_kp_person;
    this->config.overwrite_tal = config.overwrite_tal;
    this->config.use_kp_dets = config.use_kp_dets;

    this->net = readNetFromONNX(model_path);
    if (endswith(model_path, ".onnx"))
    {
        ifstream ifs("class.names");
        string line;
        while (getline(ifs, line)) this->class_names.push_back(line);
        this->num_lines = 14;
        this->num_face_pts = 0;
        // plines = (int*) crowd_segments;
    }
    else
    {
        cout << "Error: model_path: " << model_path << endl;
    }
    this->num_class = class_names.size();
}

Mat KAPAO::resize_image(Mat src_img, int* h_new, int* w_new, int* top, int* left)
{
    int src_h = src_img.rows, src_w = src_img.cols;
    *h_new = this->height;
    *w_new = this->width;
    Mat dst_img;
    if (this->keep_ratio && src_h != src_w) {
        float hw_scale = (float) src_h / src_w;
        if (hw_scale > 1)
        {
            *h_new = this->height;
            *w_new = int(this->width / hw_scale);
            resize(src_img, dst_img, Size(*w_new, *h_new), INTER_AREA);
            *left = int(this->width - *w_new) * 0.5;
            copyMakeBorder(dst_img, dst_img, 0, 0, *left, this->width - *w_new - *left, BORDER_CONSTANT, 114);
        }
        else{
            *h_new = (int)this->height * hw_scale;
            *w_new = this->width;
            resize(src_img, dst_img, Size(*w_new, *h_new), INTER_AREA);
            *top = (int)(this->height - *h_new) * 0.5;
            copyMakeBorder(dst_img, dst_img, *top, this->height - *h_new - *top, 0, 0, BORDER_CONSTANT, 114);
        }
    }
    else
    {
        resize(src_img, dst_img, Size(*w_new, *h_new), INTER_AREA);
    }
    return dst_img;
}

void KAPAO::drawPred(float conf, int left, int top, int right, int bottom, Mat& frame, int classid)
{
    // Draw a rectangle displaying the bounding box
    rectangle(frame, Point(left, top), Point(right, bottom), Scalar(0, 0, 255), 1);
    // Get the label for the class name and its confidence
    string label = format("%.2f", conf);
    label = this->class_names[classid] + ':' + label;
    // Display the label at the top of the bounding box
    int baseLine;
    Size labelSize = getTextSize(label, FONT_HERSHEY_SCRIPT_SIMPLEX, 0.5, 1, &baseLine);
    top = max(top, labelSize.height);
    rectangle(frame, Point(left, top-int(1.5*labelSize.height)),
                Point(left+int(1.5*labelSize.width), top+baseLine),
                Scalar(0, 255, 0), FILLED);
    putText(frame, label, Point(left, top), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(0, 255, 0), 1);
}

void KAPAO::detect(Mat& frame)
{
    int h_new=0, w_new=0, h_pad = 0, w_pad = 0;
    Mat dst_img = this->resize_image(frame, &h_new, &w_new, &h_pad, &w_pad);
    Mat blob = blobFromImage(dst_img, 1/255.0, Size(this->width, this->height),
                            Scalar(0, 0, 0), true, false);
    this->net.setInput(blob);
    Mat outs;
    // vector<Mat> outs;
    // this->net.forward(outs, this->net.getUnconnectedOutLayersNames());
    this->net.forward(outs);
    // Mat output = this->net.forward();

    int num_proposal = outs.size[1];   // outs.shape vector([[Batch, num_grids, 57]])
    int element_size = outs.size[2];
    if (outs.dims > 2)
    {
        outs = outs.reshape(0, num_proposal);
    }
    const int num_coords = (element_size - this->num_class - 5) * 0.5;
    // generate proposals
    vector<float> person_confidences;
    vector<Rect> person_boxes;
    vector<int> person_classIds;
    vector<float>kp_confidences;
    vector<Rect> kp_boxes;
    vector<int> kp_classIds;
    vector<vector<float> > poses;

    // Get bounding box 
    float h_ratio = (float)frame.rows / h_new, w_ratio = (float)frame.cols / w_new;
    int grid_index = 0;  // the index of grids.
    float* pdata = (float*)outs.data;   // 1 dimension array： num_grids * 57 = [element57, element57, element57, ...]
    // float * pdata = outs.ptr<float>(); 
    for (int n=0; n < this->num_stride; ++n)
    {
        const float stride = pow(2, n+3);
        int num_grid_x = (int)ceil((this->width / stride));
        int num_grid_y = (int)ceil((this->height / stride));
        for (int q = 0; q < 3; ++q)
        {
            const float w_anchor = this->anchors[n][q*2];
            const float h_anchor = this->anchors[n][q*2 + 1];
            for (int i=0; i < num_grid_y; ++i)  // grid y index
            {
                for (int j=0; j < num_grid_x; ++j)  // grid x index
                {
                    float box_score = pdata[4];
                    Mat scores = outs.row(grid_index).colRange(5, 5 + this->num_class);
                    Point classIdPoint;
                    double max_class_score;
                    // Get the value and location of the maximum score
                    minMaxLoc(scores, 0, &max_class_score, 0, &classIdPoint);
                    max_class_score *= box_score;
                    const int class_idx = classIdPoint.x;
                    
                    if  (class_idx == 0)   // person line
                    {
                        if (box_score > this->config.person_conf_thres && max_class_score > this->config.person_conf_thres)
                        {
                            float cx = (2.f * pdata[0] - 0.5f + j) * stride;  // center point x
                            float cy = (2.f * pdata[1] - 0.5f + i) * stride;  // center point y
                            float w = powf(2.f * pdata[2], 2.f) * w_anchor;  // bbox w
                            float h = powf(2.f * pdata[3], 2.f) * h_anchor;  // bbox h

                            int left = int((cx - w_pad - 0.5 * w) * w_ratio);
                            int top = int((cy - w_pad - 0.5 * h) * h_ratio);

                            person_confidences.push_back((float)max_class_score);
                            person_boxes.push_back(Rect(left, top, (int)(w*w_ratio), (int)(h*h_ratio)));
                            person_classIds.push_back(class_idx);

                            // Decode keypoints
                            vector<float> kp(3*num_coords, 0);
                            for (int k=0; k < num_coords; ++k)
                            {
                                float x = 4*pdata[5 + this->num_class + 2*k] - 2;
                                float y = 4*pdata[5 + this->num_class + 2*k + 1] - 2;
                                x *= w_anchor;
                                y *= h_anchor;
                                x += j * stride;
                                y += i * stride;
                                x = (x - w_pad) * w_ratio;
                                y = (y - h_pad) * h_ratio;
                                kp[3*k] = x;
                                kp[3*k + 1] = y;
                            }
                            poses.push_back(kp);
                        }
                    }
                    else
                    {
                        if (box_score > this->config.kp_conf_thres and max_class_score > this->config.kp_conf_thres)
                        {
                            float cx = (2.f * pdata[0] - 0.5f + j) * stride;
                            float cy = (2.f * pdata[1] - 0.5f + i) * stride;
                            float w = powf(2.f * pdata[2], 2.f) * w_anchor;
                            float h = powf(2.f * pdata[3], 2.f) * h_anchor;

                            int left = int((cx - w_pad - 0.5 * w) * w_ratio);
                            int top = int((cy - h_pad - 0.5 * h) * h_ratio);

                            kp_confidences.push_back((float)max_class_score);
                            kp_boxes.push_back(Rect(left, top, (int)(w*w_ratio), (int)(h*h_ratio)));
                            kp_classIds.push_back(class_idx);
                        }
                    }
                    grid_index++;
                    pdata += element_size;
                }
            }
        }
    }

    // Perform non-maximum suppression to eliminate redundant overlapping boxes with lower confidences.
    vector<int> person_indices;
    dnn::NMSBoxes(person_boxes, person_confidences, this->config.person_conf_thres, this->config.person_iou_thres, person_indices);
    vector<int> kp_indices;
    // TODO: all kinds of keypoint bounding boxes are NMS as the same class ?
    dnn::NMSBoxes(kp_boxes, kp_confidences, this->config.kp_conf_thres, this->config.kp_iou_thres, kp_indices);
    vector<int> pose_mask;
    for (int i = 0; i < person_indices.size(); ++i)
    {
        const int person_id = person_indices[i];
        if (person_confidences[person_id] > this->config.conf_thres_kp_person)
        {
            pose_mask.push_back(person_id);
        }
    }
    // Fuse keypoint bboxes and person keypoints
    for (int i=0; i < kp_indices.size(); ++i)
    {
        int idx = kp_indices[i];
        Rect box = kp_boxes[idx];
        float x = box.x + box.width * 0.5;
        float y = box.y + box.height * 0.5;
        float conf = kp_confidences[idx];
        int pt_id = kp_classIds[idx] - 1;
        int min_id = 0;
        float min_dist = 10000;
        for (int j=0; j < pose_mask.size(); ++j)
        {
            const int pose_id = pose_mask[j];
            const float dist = sqrt(powf(poses[pose_id][pt_id*3] - x, 2) + 
                                    powf(poses[pose_id][pt_id*3 + 1] - y, 2));
            
            if (dist < min_dist)
            {
                min_dist = dist;
                min_id = pose_id;
            }
        }
        if (conf > poses[min_id][pt_id*3 + 2] and min_dist < this->config.overwrite_tal)
        {
            poses[min_id][pt_id*3] = x;
            poses[min_id][pt_id*3 + 1] = y;
            poses[min_id][pt_id*3 + 2] = conf;
        }
    }

    // draw bbox
    for (int i=0; i < person_indices.size(); ++i)
    {
        int idx = person_indices[i];
        Rect box = person_boxes[idx];
        this->drawPred(person_confidences[idx], box.x, box.y, box.x + box.width, box.y + box.height,
                        frame, person_classIds[idx]);
    }
    // draw line
    for (int i=0; i < pose_mask.size(); ++i)
    {
        for (int j=0; j < num_coords; ++j)
        {
            if (poses[pose_mask[i]][j*3 + 2] > 0)   // if vis > 0
            {
                int x = int(poses[pose_mask[i]][j*3]);
                int y = int(poses[pose_mask[i]][j*3 + 1]);
                circle(frame, Point(x, y), 1, Scalar(0, 255, 0), -1);
            }
        }
        for (int j=0; j < this->num_lines; ++ j)
        {
            int x1 = int(poses[pose_mask[i]][this->plines[2*j]*3]);
            int y1 = int(poses[pose_mask[i]][this->plines[2*j]*3 + 1]);
            int x2 = int(poses[pose_mask[i]][this->plines[2*j + 1]*3]);
            int y2 = int(poses[pose_mask[i]][this->plines[2*j + 1]*3 + 1]);
            line(frame, Point(x1, y1), Point(x2, y2), Scalar(255, 0, 255), 1);
        }
    }

}

int main()
{
    NetConfig cfg = {0.7, 0.45, 0.5, 0.45, 0.2, 25, true};
    string img_path = "/home/huangzhiyong/Project/kapao/res/crowdpose_100024.jpg";
    std::cout << "=> Loading image from " << img_path << std::endl;
    Mat src_img = imread(img_path);
    std::cout << "=> Loading Model ..." << std::endl;
    KAPAO model(cfg, "/home/huangzhiyong/Project/kapao/cpp_detect/model.onnx");
    std::cout << "=> Detecting ... " <<  std::endl;
    model.detect(src_img);
    std::cout << "=> Done ... " <<  std::endl;

    static const string window_name = "Keypoint and Pose as Object (KAPAO)";
    namedWindow(window_name, WINDOW_NORMAL);
    imshow(window_name, src_img);
    waitKey(0);
    destroyAllWindows();
}